scimma_admin/hopskotch_auth/models.py

```diff
@@ -202,3 +207,312 @@ def rand_username(owner: User) -> str:
+        owning_group_id = topic.owning_group
+    if permission.principal == owning_group_id: # refuse to take away access from the owning group
+        return False
+    permission.delete()
+    # clean up any uses by users in the group of this permission
+    for membership in GroupMembership.objects.filter(group_id=permission.principal):
+        user_creds = SCRAMCredentials.objects.filter(owner=membership.user)
+        user_memberships = GroupMembership.objects.filter(user=membership.user)
+        user_group_perms = {}
+        for cred in user_creds:
+            permissions_to_check = CredentialKafkaPermission.objects.filter(principal=cred).select_related('parent')
+            # for each credential we must see if it has a permission which
+            # derives from the permission being removed
+            for cred_perm in permissions_to_check:
+                if cred_perm.parent!=permission:
+                    continue # great, this one is not affected by this change
+                # if affected, we need to check whether there is any other valid derivation for this 
+                # permission which could replace the one being removed
+                repairOrDeletePermission(cred_perm, 
+                                         lambda other_group_perm: other_group_perm==permission,
+                                         user_memberships, user_group_perms)
+
+    return True
+    
+
+class CredentialKafkaPermission(models.Model):
+    principal = models.ForeignKey(
+        SCRAMCredentials,
+        on_delete=models.CASCADE,
+    )
+    # individual crentials derive their permissions from group permissions, 
+    # so we track that relationship in order to coordinate changes
+    parent = models.ForeignKey(
+        GroupKafkaPermission,
+        on_delete=models.CASCADE
+    )
+    topic = models.ForeignKey(
+        KafkaTopic,
+        on_delete=models.CASCADE,
+    )
+    operation = enum.EnumField(
+        KafkaOperation,
+        editable=False,
+    )
+
+
+# This needs to be a character which is valid neither in Kafka topic names nor in enum names
+cred_perm_encoding_separator=':'
+
+
+def encode_cred_permission(parent_id, topic_id, operation) -> str:
+    return str(parent_id)+cred_perm_encoding_separator+str(topic_id)+cred_perm_encoding_separator+str(operation)
+
+
+# returns tuples of (parent ID, topic ID, operation type)
+def decode_cred_permission(encoded: str):
+    pattern = re.compile("([^"+cred_perm_encoding_separator+"]+)"+cred_perm_encoding_separator+"([^"+cred_perm_encoding_separator+"]+)"+cred_perm_encoding_separator+"([^"+cred_perm_encoding_separator+"]+)")
+    match = re.fullmatch(pattern, encoded)
+    if match is None or len(match.groups())!=3:
+        raise ValueError("Invalid encoded credential permission")
+    try:
+        op = KafkaOperation[match[3]]
+    except KeyError:
+        raise ValueError("Invalid encoded credential permission")
+    return (match[1],match[2],op)
```

Comment:
It seems simpler to use a string split. Is this equivalent? 

```python
def decode_cred_permission(encoded: str):
    parts = encoded.split(cred_perm_encoding_separator, 3)
    if len(parts) != 3:
        raise ValueError("Invalid encoded credential permission")
    try:
        op = KafkaOperation[parts[2]]
    except KeyError:
        raise ValueError("Invalid encoded credential permission")
    return (parts[0], parts[1], op)
```

Reply:
I think it is, and definitely much shorter.